librz/core/linux_heap_glibc.c

```diff
@@ -1358,48 +1362,90 @@ void GH(print_malloc_info)(RzCore *core, GHT m_state, GHT malloc_state) {
 	}
 }
 
+char *GH(rz_bin_num_to_type)(int bin_num) {
+	if (bin_num == 0) {
+		return rz_str_new("Unsorted");
+	} else if (bin_num >= 1 && bin_num <= NSMALLBINS - 1) {
+		return rz_str_new("Small");
+	} else if (bin_num >= NSMALLBINS && bin_num <= NBINS - 2) {
+		return rz_str_new("Large");
+	}
+	return NULL;
+}
+
+void GH(rz_heap_bin_free)(RzHeapBin *bin) {
+	if (!bin) {
+		return;
+	}
+	free(bin->type);
+	free(bin->message);
+	rz_list_free(bin->chunks);
+	free(bin);
+}
 /**
- * \brief Get list of chunks of <bin_num> bin from NBINS array of an arena.
+ * \brief Get information about <bin_num> bin from NBINS array of an arena.
  * \param core RzCore pointer
  * \param main_arena MallocState struct of arena
  * \param bin_num bin number of bin whose chunk list you want
- * \return RzList of RzHeapChunkListItem for the bin
+ * \return RzHeapBin struct for the bin
  */
-RZ_API RzList *GH(rz_heap_bin_content_list)(RzCore *core, MallocState *main_arena, int bin_num) {
+RZ_API RzHeapBin *GH(rz_heap_bin_content)(RzCore *core, MallocState *main_arena, int bin_num, GHT m_arena) {
 	int idx = 2 * bin_num;
 	ut64 fw = main_arena->GH(bins)[idx];
 	ut64 bk = main_arena->GH(bins)[idx + 1];
-	RzList *chunks = rz_list_newf(free);
+	RzHeapBin *bin = RZ_NEW0(RzHeapBin);
+	if (!bin) {
+		return NULL;
+	}
+	bin->fd = fw;
+	bin->bk = bk;
+	bin->bin_num = bin_num;
+	bin->type = GH(rz_bin_num_to_type)(bin_num);
+
+	// small bins hold chunks of a fixed size
+	if (!strcmp(bin->type, "Small")) {
+		bin->size = 4 * SZ + (bin_num - 1) * 2 * SZ;
+	}
+
+	bin->chunks = rz_list_newf(free);
 	GH(RzHeapChunk) *head = RZ_NEW0(GH(RzHeapChunk));
 	if (!head) {
-		return chunks;
+		return bin;
```

Comment:
Shouldn't the allocation failures in this function indicate a failure instead of returning a potentially partial bins struct?

Reply:
Yeah my bad, I will change this to returning NULL